# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [1]:
import pandas as pd
df=pd.read_csv("./e-commerce.csv")
df.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

In [2]:
df.sample(10) #随机抽出10个样本；发现数据结构完整

InvoiceNo StockCode                          Description  Quantity  \
189143    553134     23205       CHARLOTTE BAG ALPHABET  DESIGN        10   
69950     541993    72801c           4 ROSE PINK DINNER CANDLES         1   
430127    573581     21977   PACK OF 60 PINK PAISLEY CAKE CASES        24   
117886    546406       DOT                       DOTCOM POSTAGE         1   
321753    565201     22417       PACK OF 60 SPACEBOY CAKE CASES        24   
281202    561515     21977   PACK OF 60 PINK PAISLEY CAKE CASES         1   
402313    571486     21876                        POTTERING MUG        12   
422466    573085     21623         VINTAGE UNION JACK MEMOBOARD         2   
330311    565917     23209             LUNCH BAG VINTAGE DOILY          2   
26887     538524     21640  ASSORTED TUTTI FRUTTI  FOB NOTEBOOK         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
189143   5/13/2011 12:50       0.85     15382.0  United Kingdom  
69950     1/25/2011 9:37       2.46         NaN  United Kingdom  
430127  10/31/2011 14:22       0.55     14911.0            EIRE  
117886   3/11/2011 16:21     201.37         NaN  United Kingdom  
321753    9/1/2011 16:41       0.55     14907.0  United Kingdom  
281202   7/27/2011 15:16       1.25         NaN  United Kingdom  
402313  10/17/2011 14:59       1.65     17612.0  United Kingdom  
422466  10/27/2011 14:27       9.95     14441.0  United Kingdom  
330311    9/7/2011 16:15       4.13         NaN  United Kingdom  
26887    12/13/2010 9:35       1.66         NaN  United Kingdom

In [3]:
df.info() #查看数据的基本信息。1.有无空值:发现Description，CustomerID存在空值 2.Dtype 类型里面:InvoiceDate应该是时间 datetime64;CustomerID应该是字符串 需要进行类型的转换


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


这个 DataFrame 有 541,909 行，8 列，列出了每一列的名称、数据类型、非空值数量，以及内存占用情况。下面是输出内容的详细说明：
解释：
RangeIndex: RangeIndex: 541909 entries, 0 to 541908数据集共有 541,909 行，索引从 0 到 541,908。
Data columns (total 8 columns):数据集中共有 8 列，每列都有自己的名称和数据类型。
每列信息：
InvoiceNo (obj|ect):发票号列，数据类型为 object（通常是字符串），包含 541,909 个非空值，说明该列没有缺失值。
StockCode (object):商品代码列，数据类型为 object，也包含 541,909 个非空值，表示没有缺失值。
Description (object):商品描述列，数据类型为 object，包含 540,455 个非空值，表示有部分缺失值（541,909 - 540,455 = 1,454 个缺失值）。
Quantity (int64):购买数量列，数据类型为 int64，包含 541,909 个非空值，没有缺失值，且所有值为整数。
InvoiceDate (object):发票日期列，数据类型为 object，包含 541,909 个非空值，表示日期是以字符串形式存储的。
UnitPrice (float64):单价列，数据类型为 float64，包含 541,909 个非空值，表示价格是以浮点数存储的。
CustomerID (float64):客户ID列，数据类型为 float64，包含 406,829 个非空值，表示有部分缺失值（541,909 - 406,829 = 135,080 个缺失值）。
Country (object):国家列，数据类型为 object，包含 541,909 个非空值，没有缺失值，说明每个交易都有国家信息。
数据类型：
float64：表示浮点数的列，共有 2 列（UnitPrice 和 CustomerID）。
int64：表示整数的列，只有 1 列（Quantity）。
object：表示字符串（或者混合类型）的列，共有 5 列（InvoiceNo、StockCode、Description、InvoiceDate、Country）。




评估缺失数据

In [4]:
df["Description"].isnull().sum()

1454

In [5]:
df[df["Description"].isnull()] #提取空值行

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

我们可以发现 Description为NaN的，UnitPrice为0？ 验证猜想

In [6]:
df[(df["Description"].isnull()&df["UnitPrice"] !=0)] #df[布尔值] 查询Description为NaN的，UnitPrice不为0？的值，发现 NaN值的UnitePrice均为0即对我们的分析无帮助。
#同时缺失可以进行删除.[描述和单价是重要的分析变量]；后面可以删除

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [7]:
df["UnitPrice"] !=0

0         True
1         True
2         True
3         True
4         True
          ... 
541904    True
541905    True
541906    True
541907    True
541908    True
Name: UnitPrice, Length: 541909, dtype: bool

In [8]:
df[df["UnitPrice"]!=0]

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[539394 rows x 8 columns]

CustomerID也有空值

In [9]:
df["CustomerID"].isnull()

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Name: CustomerID, Length: 541909, dtype: bool

In [10]:
df[df["CustomerID"].isnull()] # ID缺失不影响分析


InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

评估重复或者不一致数据

不一致数据可能存在与Country里面，是否不同值指代一个意思。InvoiceNo	StockCode CustomerID均可以重复 但是不能同时出现重复

In [11]:
df["Country"].value_counts() #value_counts() 是 Pandas 中非常有用的方法，用于统计每个唯一值的出现次数。
#它返回一个 Series，索引是唯一值，值是出现的频次。通常用于对单列进行频率统计
#可以发现：United Kingdom ；UK；U.K. 重复 ;United States  USA

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


In [12]:
df[df.duplicated()] #存在5268 条完全相同的重复行

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517     12/1/2010 11:45       1.25     17908.0  United Kingdom  
527     12/1/2010 11:45       2.10     17908.0  United Kingdom  
537     12/1/2010 11:45       2.95     17908.0  United Kingdom  
539     12/1/2010 11:45       4.95     17908.0  United Kingdom  
555     12/1/2010 11:49       2.95     17920.0  United Kingdom  
...                 ...        ...         ...             ...  
541675  12/9/2011 11:34       0.39     14446.0  United Kingdom  
541689  12/9/2011 11:34       2.49     14446.0  United Kingdom  
541692  12/9/2011 11:34       1.95     14446.0  United Kingdom  
541699  12/9/2011 11:34       2.10     14446.0  United Kingdom  
541701  12/9/2011 11:34       2.08     14446.0  United Kingdom  

[5268 rows x 8 columns]

In [13]:
df.duplicated().sum()

5268

In [14]:
df["Country"].replace(["United Kingdom","U.K."],"UK") #采用replace进行值的替换

0             UK
1             UK
2             UK
3             UK
4             UK
           ...  
541904    France
541905    France
541906    France
541907    France
541908    France
Name: Country, Length: 541909, dtype: object

In [15]:
df["Country"].replace(["United States"]," USA")

0         United Kingdom
1         United Kingdom
2         United Kingdom
3         United Kingdom
4         United Kingdom
               ...      
541904            France
541905            France
541906            France
541907            France
541908            France
Name: Country, Length: 541909, dtype: object

In [16]:
df["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


In [17]:
df["CustomerID"]=df["CustomerID"].astype(str)#进行类型转换并且赋值

评估无效和错误数据

In [18]:
df.describe() #发现最小值存在着复数; CID 应当为object 不应该进行数值计算

Quantity      UnitPrice
count  541909.000000  541909.000000
mean        9.552250       4.611114
std       218.081158      96.759853
min    -80995.000000  -11062.060000
25%         1.000000       1.250000
50%         3.000000       2.080000
75%        10.000000       4.130000
max     80995.000000   38970.000000

In [19]:
df[df["Quantity"]<0]#如果这个代码以字母“c”开头，表示这笔交易被取消。 思考：是不是所有取消的订单，数量均为负数呢？

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice CustomerID         Country  
141      12/1/2010 9:41      27.50    14527.0  United Kingdom  
154      12/1/2010 9:49       4.65    15311.0  United Kingdom  
235     12/1/2010 10:24       1.65    17548.0  United Kingdom  
236     12/1/2010 10:24       0.29    17548.0  United Kingdom  
237     12/1/2010 10:24       0.29    17548.0  United Kingdom  
...                 ...        ...        ...             ...  
540449   12/9/2011 9:57       0.83    14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69    15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95    15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25    17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25    17315.0  United Kingdom  

[10624 rows x 8 columns]

In [20]:
df["InvoiceNo"].str[0]=="C" #==C 不能做赋值操作

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Name: InvoiceNo, Length: 541909, dtype: bool

In [21]:
df[df["InvoiceNo"].str[0]=="C"]#提取出以C开头的

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice CustomerID         Country  
141      12/1/2010 9:41      27.50    14527.0  United Kingdom  
154      12/1/2010 9:49       4.65    15311.0  United Kingdom  
235     12/1/2010 10:24       1.65    17548.0  United Kingdom  
236     12/1/2010 10:24       0.29    17548.0  United Kingdom  
237     12/1/2010 10:24       0.29    17548.0  United Kingdom  
...                 ...        ...        ...             ...  
540449   12/9/2011 9:57       0.83    14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69    15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95    15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25    17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25    17315.0  United Kingdom  

[9288 rows x 8 columns]

In [22]:
df[(df["InvoiceNo"].str[0]!="C")&(df["Quantity"]<0)]#假设：不以C开头并且 数量小于零存在， 异常值单价 全为0？

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

       CustomerID         Country  
2406          nan  United Kingdom  
4347          nan  United Kingdom  
7188          nan  United Kingdom  
7189          nan  United Kingdom  
7190          nan  United Kingdom  
...           ...             ...  
535333        nan  United Kingdom  
535335        nan  United Kingdom  
535336        nan  United Kingdom  
536908        nan  United Kingdom  
538919        nan  United Kingdom  

[1336 rows x 8 columns]

In [23]:
df[(df["InvoiceNo"].str[0]!="C")&(df["Quantity"]<0)&(df["UnitPrice"]!=0)]# 结论：不以C开头并且数量小于0的单价全为0 ，可以删除

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [24]:
#单价也有负数
df[df["UnitPrice"]<0] #可以看到 描述是坏账处理 不属于实际交易商品 后续可以删除

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice CustomerID         Country  
299983  -11062.06        nan  United Kingdom  
299984  -11062.06        nan  United Kingdom

## 清理数据

根据前面的评估部分得到的结论，需要对数据清理包括：
把InvoiceDate变量数据类型转换为日期时间
把CustomerID变量数据转化为字符串类型
把Desciption变量缺失值删除
把Country变量值USA替换为United States;UK U.K 替换为United Kindom
把quntity变量值为负数的删除
把unitprice变量值为负数的观察值删除


 为了区分清理后的数据和原始数据，创建新变量clean_df，让他为df复制出来的副本，之后所有的操作均再clean_df进行

In [25]:
clean_df=df.copy()
clean_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice CustomerID         Country  
0        12/1/2010 8:26       2.55    17850.0  United Kingdom  
1        12/1/2010 8:26       3.39    17850.0  United Kingdom  
2        12/1/2010 8:26       2.75    17850.0  United Kingdom  
3        12/1/2010 8:26       3.39    17850.0  United Kingdom  
4        12/1/2010 8:26       3.39    17850.0  United Kingdom  
...                 ...        ...        ...             ...  
541904  12/9/2011 12:50       0.85    12680.0          France  
541905  12/9/2011 12:50       2.10    12680.0          France  
541906  12/9/2011 12:50       4.15    12680.0          France  
541907  12/9/2011 12:50       4.15    12680.0          France  
541908  12/9/2011 12:50       4.95    12680.0          France  

[541909 rows x 8 columns]

In [26]:
clean_df["InvoiceDate"]=pd.to_datetime(clean_df["InvoiceDate"])#时间格式更改pd.to_datetime 然后进行赋值

In [27]:
#df["CustomerID"]=df["CustomerID"].astype("int") #这是错误的,customerID 里面包含空值
#df.info() 查看后，发现确实已经更改，但是我们需要删除0


In [28]:
clean_df["CustomerID"]=clean_df["CustomerID"].astype(str)
clean_df#发现CID 含有 0 用str的slice方法

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55    17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39    17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75    17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39    17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39    17850.0  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85    12680.0          France  
541905 2011-12-09 12:50:00       2.10    12680.0          France  
541906 2011-12-09 12:50:00       4.15    12680.0          France  
541907 2011-12-09 12:50:00       4.15    12680.0          France  
541908 2011-12-09 12:50:00       4.95    12680.0          France  

[541909 rows x 8 columns]

In [29]:
clean_df["CustomerID"].str.slice(0,-2)
clean_df["CustomerID"]=df["CustomerID"].str.slice(0,-2)

In [30]:
clean_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[541909 rows x 8 columns]

In [31]:
clean_df.dropna(subset="Description",inplace=True)

In [32]:
clean_df["Description"].isnull().sum()

0

替换值 ：replace 


In [33]:
clean_df=clean_df.replace(["UK","U.K."],"United Kingdom")



In [34]:
clean_df["Country"]=clean_df["Country"].replace("USA","United States")

In [35]:
clean_df["Country"].value_counts()

Country
United Kingdom          494024
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
United States              291
China                      288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


In [36]:
print(len(clean_df[clean_df["Country"]=="USA"]))#检查 country为USA行数的长度
print(len(clean_df[clean_df["Country"]=="UK"]))
print(len(clean_df[clean_df["Country"]=="U.K"]))

0
0
0


Quantity<0 & Unit Price<0 的值全部删除 思路：保留 数量大于0以及 单价大于0的行数

In [37]:
clean_df["Quantity"]=clean_df["Quantity"][clean_df["Quantity"]>0]
clean_df["UnitPrice"]=clean_df["UnitPrice"][clean_df["UnitPrice"]>0]

In [38]:
print(len(clean_df[clean_df["Quantity"]<0]))
print(len(clean_df[clean_df["UnitPrice"]<0]))

0
0


## 保存清理后的数据

In [39]:
clean_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER       6.0   
1         536365     71053                  WHITE METAL LANTERN       6.0   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER       8.0   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS      12.0   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL        6.0   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL        4.0   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE       4.0   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT        3.0   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[540455 rows x 8 columns]

In [42]:

clean_df.to_csv("e_commerce_clean_df.csv",index=False) #会默认保留第一列索引 index=False 不保留行索引

In [44]:
pd.read_csv("e_commerce_clean_df.csv")

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER       6.0   
1         536365     71053                  WHITE METAL LANTERN       6.0   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER       8.0   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
...          ...       ...                                  ...       ...   
540450    581587     22613          PACK OF 20 SPACEBOY NAPKINS      12.0   
540451    581587     22899         CHILDREN'S APRON DOLLY GIRL        6.0   
540452    581587     23254        CHILDRENS CUTLERY DOLLY GIRL        4.0   
540453    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE       4.0   
540454    581587     22138        BAKING SET 9 PIECE RETROSPOT        3.0   

                InvoiceDate  UnitPrice CustomerID         Country  
0       2010-12-01 08:26:00       2.55      17850  United Kingdom  
1       2010-12-01 08:26:00       3.39      17850  United Kingdom  
2       2010-12-01 08:26:00       2.75      17850  United Kingdom  
3       2010-12-01 08:26:00       3.39      17850  United Kingdom  
4       2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                     ...        ...        ...             ...  
540450  2011-12-09 12:50:00       0.85      12680          France  
540451  2011-12-09 12:50:00       2.10      12680          France  
540452  2011-12-09 12:50:00       4.15      12680          France  
540453  2011-12-09 12:50:00       4.15      12680          France  
540454  2011-12-09 12:50:00       4.95      12680          France  

[540455 rows x 8 columns]